File: Server~/src/tools/renameAssetTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'rename_asset';
const toolDescription = `Renames an asset in place (changes filename only, not directory), handling .meta files automatically.

Identify the asset by assetPath and/or guid (at least one required).
The newName parameter should be just the filename without path or extension (e.g. "PlayerModel" not "Assets/PlayerModel.fbx").
If you need to move an asset to a different folder, use move_asset instead.`;

const paramsSchema = z.object({
  assetPath: z.string().optional().describe('Current asset path (e.g. "Assets/Models/OldName.fbx")'),
  guid: z.string().optional().describe('Asset GUID'),
  newName: z.string().describe('New filename without path or extension (e.g. "NewName")')
});

export function registerRenameAssetTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);
  server.tool(toolName, toolDescription, paramsSchema.shape, async (params: any) => {
    try {
      logger.info(`Executing tool: ${toolName}`, params);
      const result = await toolHandler(mcpUnity, params);
      logger.info(`Tool execution successful: ${toolName}`);
      return result;
    } catch (error) {
      logger.error(`Tool execution failed: ${toolName}`, error);
      throw error;
    }
  });
}

async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  if ((!params.assetPath || params.assetPath.trim() === '') && (!params.guid || params.guid.trim() === '')) {
    throw new McpUnityError(ErrorType.VALIDATION, "At least one of 'assetPath' or 'guid' must be provided");
  }
  if (!params.newName || params.newName.trim() === '') {
    throw new McpUnityError(ErrorType.VALIDATION, "Required parameter 'newName' must be provided");
  }

  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: { assetPath: params.assetPath, guid: params.guid, newName: params.newName }
  });

  if (!response.success) {
    throw new McpUnityError(ErrorType.TOOL_EXECUTION, response.message || 'Failed to rename asset');
  }

  let text = response.message || 'Asset renamed successfully';
  if (response.data) {
    text += `\n\nPrevious path: ${response.data.previousPath}`;
    text += `\nNew path: ${response.data.assetPath}`;
    text += `\nGUID: ${response.data.guid}`;
  }

  return { content: [{ type: response.type, text }] };
}

File: Editor/Tools/RenameAssetTool.cs
using System;
using McpUnity.Unity;
using McpUnity.Utils;
using UnityEditor;
using Newtonsoft.Json.Linq;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for renaming an asset in place via AssetDatabase
    /// </summary>
    public class RenameAssetTool : McpToolBase
    {
        public RenameAssetTool()
        {
            Name = "rename_asset";
            Description = "Renames an asset in place (filename only, not path), handling .meta files automatically";
        }

        public override JObject Execute(JObject parameters)
        {
            string assetPath = parameters["assetPath"]?.ToObject<string>()?.Trim();
            string guid = parameters["guid"]?.ToObject<string>()?.Trim();
            string newName = parameters["newName"]?.ToObject<string>()?.Trim();

            // Resolve source asset
            string resolvedPath = MoveAssetTool.ResolveAssetPath(assetPath, guid, out _, out JObject error);
            if (error != null) return error;

            // Validate new name
            if (string.IsNullOrEmpty(newName))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'newName' not provided",
                    "validation_error"
                );
            }

            // newName should be just a name, not a path
            if (newName.Contains("/") || newName.Contains("\\"))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Parameter 'newName' must be a filename only (no path separators). Use move_asset to change directories.",
                    "validation_error"
                );
            }

            if (newName.Contains(".."))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Parameter 'newName' must not contain '..'",
                    "validation_error"
                );
            }

            // Strip extension if provided — RenameAsset expects name without extension
            string currentExtension = System.IO.Path.GetExtension(resolvedPath);
            if (!string.IsNullOrEmpty(currentExtension) && newName.EndsWith(currentExtension, StringComparison.OrdinalIgnoreCase))
            {
                newName = newName.Substring(0, newName.Length - currentExtension.Length);
            }

            if (string.IsNullOrWhiteSpace(newName))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "New name cannot be empty after removing extension",
                    "validation_error"
                );
            }

            try
            {
                string result = AssetDatabase.RenameAsset(resolvedPath, newName);
                if (!string.IsNullOrEmpty(result))
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Failed to rename asset: {result}",
                        "rename_error"
                    );
                }

                // Build the new path
                string directory = System.IO.Path.GetDirectoryName(resolvedPath)?.Replace("\\", "/");
                string newPath = $"{directory}/{newName}{currentExtension}";
                string newGuid = AssetDatabase.AssetPathToGUID(newPath);

                McpLogger.LogInfo($"[MCP Unity] Renamed asset from '{resolvedPath}' to '{newPath}'");

                return new JObject
                {
                    ["success"] = true,
                    ["type"] = "text",
                    ["message"] = $"Successfully renamed asset from '{System.IO.Path.GetFileName(resolvedPath)}' to '{newName}{currentExtension}'",
                    ["data"] = new JObject
                    {
                        ["previousPath"] = resolvedPath,
                        ["assetPath"] = newPath,
                        ["guid"] = newGuid
                    }
                };
            }
            catch (Exception ex)
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    $"Error renaming asset: {ex.Message}",
                    "rename_error"
                );
            }
        }
    }
}
